File: iOS 8 by Tutorials/Objective-C/Chapter 13 - Extensions - Document Providers/Imgvue/Imgvue/S3Images/DocumentPickerViewController.m
/*
 * Copyright (c) 2014 Razeware LLC
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

@import UIKit;
@import UIWidgets;
@import S3Kit;

#import "DocumentPickerViewController.h"
#import "UploadViewController.h"

@interface DocumentPickerViewController () <ThumbnailDelegate, NSFileManagerDelegate>

@end

@implementation DocumentPickerViewController

- (void)didSelectImage:(S3Image *)image {
  NSURL* imageUrl = image.url;
  NSString* filename = [imageUrl.pathComponents lastObject];
  NSURL* outUrl = [self.documentStorageURL URLByAppendingPathComponent:filename];
  
  NSFileCoordinator* coordinator = [[NSFileCoordinator alloc] init];
  [coordinator coordinateWritingItemAtURL:outUrl
                                  options:NSFileCoordinatorWritingForReplacing
                                    error:nil
                               byAccessor:^(NSURL *newURL) {
                                 NSFileManager* fm = [[NSFileManager alloc] init];
                                 fm.delegate = self;
                                 [fm copyItemAtURL:imageUrl toURL:outUrl error:nil];
                               }];

  [self dismissGrantingAccessToURL:outUrl];
}

- (BOOL)fileManager:(NSFileManager *)fileManager shouldProceedAfterError:(NSError *)error copyingItemAtURL:(NSURL *)srcURL toURL:(NSURL *)dstURL {
  return YES;
}

- (void)prepareForPresentationInMode:(UIDocumentPickerMode)mode {
  UINavigationController* nav = self.childViewControllers[0];
  ThumbnailCollectionViewController* thumbController = (ThumbnailCollectionViewController*) [nav topViewController];
  
  [thumbController.navigationController setNavigationBarHidden:YES animated:NO];
  
  if (mode == UIDocumentPickerModeExportToService || mode == UIDocumentPickerModeMoveToService) {
    [thumbController performSegueWithIdentifier:@"upload" sender:nil];
  } else {
    thumbController.delegate = self;
  }
}

- (void)nameChosen:(NSString *)name {
  NSString* justFilename = [name stringByDeletingPathExtension];
  NSURL* exportURL = [[self.documentStorageURL URLByAppendingPathComponent:justFilename] URLByAppendingPathExtension:@"jpg"];
  [self upload:name url:exportURL];
}

- (void) upload:(NSString*)name url:(NSURL*)outURL {
  BOOL access = [self.originalURL startAccessingSecurityScopedResource];
  if (access) {
    NSFileCoordinator* fc = [[NSFileCoordinator alloc] init];
    NSError* error = nil;
    [fc coordinateReadingItemAtURL:self.originalURL
                           options:NSFileCoordinatorReadingForUploading
                             error:&error byAccessor:^(NSURL *newURL) {
                               BucketInfo* bucket = [[BucketInfo alloc] init];
                               NSData* data = [NSData dataWithContentsOfURL:newURL];
                               [bucket uploadData:data name:name completion:^(NSError *error) {
                                 if (error) {
                                   NSLog(@"error uploading %@", error);
                                 }
                               }];
                               [data writeToURL:outURL atomically:YES];
                               [self dismissGrantingAccessToURL:outURL];
                             }];
    [self.originalURL stopAccessingSecurityScopedResource];
  }
}


@end

File: iOS 8 by Tutorials/Swift/Chapter 13 - Extensions - Document Providers/Imgvue-Final/Imgvue/S3Images/DocumentPickerViewController.swift
/*
* Copyright (c) 2014 Razeware LLC
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
* THE SOFTWARE.
*/

import UIKit
import UIWidgets
import S3Kit

class DocumentPickerViewController: UIDocumentPickerExtensionViewController, ThumbnailDelegate, NSFileManagerDelegate, UploadControllerDelegate {

  override func prepareForPresentationInMode(mode: UIDocumentPickerMode) {
    super.prepareForPresentationInMode(mode)
    
    let vc = self.childViewControllers;
    let nav = vc.first as UINavigationController
    
    let thumbController = nav.topViewController as ThumbnailCollectionViewController
    
    thumbController.navigationController?.setNavigationBarHidden(true, animated: false)
    if mode == .ExportToService || mode == .MoveToService {
      thumbController.performSegueWithIdentifier("upload", sender: nil)
    } else {
      thumbController.delegate = self
    }
  }
  
  // MARK: ThumbnailDelegate
  func didSelectImage(image: S3Image) {
    let imageUrl = image.url!
    
    //1
    let filename = imageUrl.pathComponents.last as String
    
    //2
    let outUrl = documentStorageURL.URLByAppendingPathComponent(filename)
    
    // 3
    let coordinator = NSFileCoordinator()
    coordinator.coordinateWritingItemAtURL(outUrl,
                                           options: .ForReplacing,
                                           error: nil,
                                           byAccessor: { newURL in
      //4
      let fm = NSFileManager()
      fm.delegate = self
      fm.copyItemAtURL(imageUrl, toURL: newURL, error: nil)
    })
    
    // 5
    dismissGrantingAccessToURL(outUrl)
  }
  
  // MARK: NSFileManagerDelegate
  func fileManager(fileManager: NSFileManager,
      shouldProceedAfterError error: NSError,
      copyingItemAtURL srcURL: NSURL,
      toURL dstURL: NSURL) -> Bool {
      
      return true
  }

  // MARK: UploadControllerDelegate
  func nameChosen(name: String) {
    let justFilename = name.stringByDeletingPathExtension
    let exportURL = documentStorageURL.URLByAppendingPathComponent(name).URLByAppendingPathExtension("jpg")
    upload(name, outURL: exportURL)
  }

  func upload(name: String, outURL: NSURL) {
    let access = originalURL!.startAccessingSecurityScopedResource() //1
    if access {
      let fc = NSFileCoordinator()
      var error: NSError?
      fc.coordinateReadingItemAtURL(originalURL!, //2
        options: .ForUploading,
        error: &error) { url in
          let bucket = BucketInfo() //3
          let data = NSData(contentsOfURL: url)!
          bucket.uploadData(data, name: name) { error in //4
            if error != nil {
              println("error uploading: \(error)")
            }
          }
          data.writeToURL(outURL, atomically: true) //5
          self.dismissGrantingAccessToURL(outURL) //6
      }
      originalURL!.stopAccessingSecurityScopedResource()
    }
  }
}
